src/libertem/udf/com.py

```diff
@@ -0,0 +1,630 @@
+
+            Please note that a regression calculated on the data itself may
+            distort valid results, such as the effect of long-range fields.
+            Furthermore, it may yield erronerous results due to aliasing on
+            periodic structures or boundary effects. Subtracting a reference
+            measurement or regression on a reference measurement instead of a
+            regression on the data may give more reliable results.
+        """
+        if ri >= r:
+            raise ValueError('Inner radius must be less than outer radius for annular CoM')
+        return cls(
+            com_params=CoMParams(
+                cy=cy, cx=cx, r=r, ri=ri,
+                scan_rotation=scan_rotation, flip_y=flip_y,
+                regression=regression,
+            )
+        )
+
+    def get_backends(self):
+        return self.BACKEND_ALL
+
+    def get_result_buffers(self):
+        dtype = np.result_type(self.meta.input_dtype, np.float32)
+        return {
+            'raw_mask_result': self.buffer(
+                kind='nav', dtype=dtype, extra_shape=(3, ), where='device', use='private'
+            ),
+            'raw_com': self.buffer(
+                kind='nav', dtype=dtype, extra_shape=(2, ), use='result_only'
+            ),
+            'raw_shifts': self.buffer(
+                kind='nav', dtype=dtype, extra_shape=(2, ), use='result_only'
+            ),
+            'field': self.buffer(
+                kind='nav', dtype=dtype, extra_shape=(2, ), use='result_only'
+            ),
+            'magnitude': self.buffer(
+                kind='nav', dtype=dtype, use='result_only'
+            ),
+            'divergence': self.buffer(
+                kind='nav', dtype=dtype, use='result_only'
+            ),
+            'curl': self.buffer(
+                kind='nav', dtype=dtype, use='result_only'
+            ),
+            'regression': self.buffer(
+                kind='single', extra_shape=(3, 2), dtype=np.float64, use='result_only'
+            ),
+        }
+
+    def get_params(self) -> CoMParams:
+        sig_shape = tuple(self.meta.dataset_shape.sig)
+        cy = self.params.com_params.cy
+        if cy is None:
+            cy = sig_shape[0] // 2
+
+        cx = self.params.com_params.cx
+        if cx is None:
+            cx = sig_shape[1] // 2
+
+        r = self.params.com_params.r
+        ri = self.params.com_params.ri
+        scan_rotation = self.params.com_params.scan_rotation
+        flip_y = self.params.com_params.flip_y
+        regression = self.params.com_params.regression
+
+        cp = CoMParams(
+            cy=cy, cx=cx, r=r, ri=ri, scan_rotation=scan_rotation, flip_y=flip_y,
+            regression=regression,
+        )
+        return cp
+
+    def get_task_data(self):
+        sig_shape = tuple(self.meta.dataset_shape.sig)
+        com_params = self.get_params()
+        if len(sig_shape) != 2:
+            raise ValueError('CoMUDF only works with 2D sig shape.')
```

Comment:
Should we have test cases for incompatible sig/nav shapes? Should be easy enough with a memory dataset.